### ISSUE TITLE
Servertime in the future breaks check results processing

### ISSUE THREAD

**Issue by marcofl** [Icinga/icinga2#6797]
## Expected Behavior
When a server time on a client was in the future and then the time is fixed, check results should be processed again normally on the master. 

## Current Behavior
HW issues, users setting the server time wrong manually, etc. can lead to a server time in the future  (say 1 month in the future). We than fix the time and make sure NTP works correctly again. After that, icinga2 on the master is not accepting / processing check results from the client again. 

In icingaweb2 you'll see last check with a date in the future and next check will show negative values. This will not change until the date in the future is reached. 

## Possible Solution
I don't know how time differences for check results are currently handled, but I would think something like this should be the case:
* check result times are in UTC
* when the master receives a check result and the time of a check result is in the future compared to masters time it ignores the result and throws a warning  

## Steps to Reproduce (for bugs)
1. Have a master (-sat) - client icinga2 cluster setup. on a client set the time on the server to something in the future (e.g. 1 hour)
2. you'll see check results on the master in icingaweb2 that show the date in the future for "last check"
3. then fix the time on the client
4. you'll still see the checks with "last check" in the future and you'll not get any fresh check results for all the servers that run (check source) on that client
![screen shot 2018-11-23 at 11 15 47](https://user-images.githubusercontent.com/8982180/48940053-97ce5180-ef16-11e8-8814-9df166e280f7.png)

## Context
A user set the server time to something 1 month in the future to test something. Icinga2 was running as a client on this server and continues to send check results. Our NTP check of that sever complained about the time difference. that way we knew something is broken. after fixing the time, we  were not getting any fresh results for all the check that run on that client (check source).

Now it's impossible to get new results for the services on that client. I tried deleting the host and services from the icinga2 config and readd them after another reload. Also removed the host and services from the IDO DB manually - now the checks are pending and not showing any check results in icingaweb2. 

## Your Environment

* Version used (`icinga2 --version`):
```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.10.2-1)

Copyright (c) 2012-2018 Icinga Development Team (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Ubuntu
  Platform version: 16.04.5 LTS (Xenial Xerus)
  Kernel: Linux
  Kernel version: 4.4.0-139-generic
  Architecture: x86_64

Build information:
  Compiler: GNU 5.3.1
  Build host: 409e1113863b

Application information:

General paths:
  Config directory: /etc/icinga2
  Data directory: /var/lib/icinga2
  Log directory: /var/log/icinga2
  Cache directory: /var/cache/icinga2
  Spool directory: /var/spool/icinga2
  Run directory: /run/icinga2

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid
```
* Operating System and version: Ubuntu 16.04
* Enabled features (`icinga2 feature list`): 
```
Disabled features: compatlog debuglog elasticsearch graphite livestatus opentsdb perfdata statusdata syslog
Enabled features: api checker command gelf ido-mysql influxdb mainlog notification
```
* Icinga Web 2 version and modules (System - About): 2.6.1-1.xenial
* Config validation (`icinga2 daemon -C`): not relevant
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

**Comment by Crunsher** [comment]
>it should heal itself after some time

I'm not sure Icinga 2 can tell for sure if something like that went wrong, I'd be concerned about false positives. That's why resolving this on a user request makes more sense to me.

>That would already be good enough for me. Do you know of any workaround which is not yet documented? I tried a lot and could not get it working again for that client.

The problem is with the statefile. Simplest way would be to just delete it, but then you lose state - ie. current state, acknowledgments etc. It should be possible to heal the statefile, but I can't give a step-by-step guide without reproducing this first.

**Comment by marcofl** [comment]
As this "time travel" situation usually does not occur too often, I guess a manual fix would be okay. Something like "editing" the state file could help, is there a editor for the state file or anything in this direction? I mean deleting one host object with all its service or all objects with a specific check source would be enough I guess.

Deleting the sate file completely is not really an option for us, this would cause a lot of issue with acked, services, re-notification, etc.

**Comment by dnsmichi** [comment]
The state file uses JSON wrapped into Netstring. I guess one can write a script to modify it and generate a safe netstring again. I wouldn't want to officially support such a thing through a CLI command though.

**Comment by marcofl** [comment]
sounds like a experiment, what could possibly go wrong, I mean fortunately it's not CSV :D

**Comment by marcofl** [comment]
> The state file uses JSON wrapped into Netstring. I guess one can write a script to modify it and generate a safe netstring again. I wouldn't want to officially support such a thing through a CLI command though.

what came into my mind: when does icinga2 purge objects from the statefile when they do no longer exist in the config? because removing the host / services from the config and then restarting icinga2 would fix my issue - but it does not.

**Comment by Napsty** [comment]
This happened to me today as well. I installed a new motherboard and CPU on my testserver on which an Icinga2 instance runs (2.10.2-1.stretch). For some reason the motherboard's date was one day in the future, December 16 2018. So the server booted up with the new hardware, being a day in the future. Icinga2 started up and did some checks, thinking it is in the future. 

This can be verified in icingaweb2:

![image](https://user-images.githubusercontent.com/3702203/50042585-648a6880-0065-11e9-9666-b8b6564879bd.png)

I can also see the graphs written into InfluxDB in the future:

![image](https://user-images.githubusercontent.com/3702203/50042617-c77bff80-0065-11e9-9022-e7c02b53fd06.png)

After a reboot of the server, Linux synced the time to hwclock. When the server was back up, it was running in the present again. But, as @marcofl said, checks are not executed anymore, probably until the date and time in the future is present (back to the future, right? haha!). 
Even a forced re-check will not run.

So far reloads and restarts do not help.

**Comment by Napsty** [comment]
Here is some additional information from one of the affected checks. Maybe it gives some clarification. 
Note current (correct) timestamp on the Icinga2 server is:

```
root@icinga2:~# date +%s
1544874238
 ```

By analyzing the entry of this relevant service in the statefile (using the debugging script from @lazyfrosch -> https://gist.github.com/lazyfrosch/f05b5cbfe8a05d4a1cee93db2c0028c0) you can see the timestamps in the future, `"execution_end": 1544930411.889406,` for example.

```  
    {
        "name": "bw-nas!Diskspace /srv/data",
        "type": "Service",
        "update": {
            "acknowledgement": 0.0,
            "acknowledgement_expiry": 0.0,
            "check_attempt": 1.0,
            "flapping": false,
            "flapping_buffer": 262144.0,
            "flapping_current": 5.9,
            "flapping_index": 19.0,
            "flapping_last_change": 0.0,
            "force_next_check": true,
            "force_next_notification": false,
            "last_check_result": {
                "active": true,
                "check_source": "napsty.example.com",
                "command": [
                    "/usr/lib/nagios/plugins/check_nrpe",
                    "-H",
                    "192.168.15.50",
                    "-c",
                    "check_disk",
                    "-a",
                    "3%",
                    "1%",
                    "/srv/data",
                    "10%",
                    "5%"
                ],
                "execution_end": 1544930411.889406,
                "execution_start": 1544930408.802476,
                "exit_status": 2.0,
                "output": "connect to address 192.168.15.50 port 5666: No route to host\nconnect to host 192.168.15.50 port 5666: No route to host",
                "performance_data": [],
                "schedule_end": 1544930411.889547,
                "schedule_start": 1544930408.8019118,
                "state": 2.0,
                "ttl": 0.0,
                "type": "CheckResult",
                "vars_after": {
                    "attempt": 1.0,
                    "reachable": false,
                    "state": 2.0,
                    "state_type": 0.0
                },
                "vars_before": {
                    "attempt": 1.0,
                    "reachable": true,
                    "state": 0.0,
                    "state_type": 1.0
                }
            },
            "last_hard_state_change": 1544423091.744999,
            "last_hard_state_raw": 0.0,
            "last_reachable": true,
            "last_state_change": 1544930411.889547,
            "last_state_critical": 1544930411.88957,
            "last_state_ok": 1544799404.391699,
            "last_state_raw": 2.0,
            "last_state_type": 1.0,
            "last_state_unknown": 0.0,
            "last_state_unreachable": 1544930411.889574,
            "last_state_warning": 1544422796.716093,
            "next_check": 1544876271.0,
            "state_raw": 2.0,
            "state_type": 0.0,
            "type": "Service",
            "version": 0.0
        }
    },
```

**Comment by marcofl** [comment]
thanks for the link to that script @Napsty :-)
for me it looks a bit different:
```
    {
        "name": "server1.example.com!load",
        "type": "Service",
        "update": {
            "acknowledgement": 0.0,
            "acknowledgement_expiry": 0.0,
            "check_attempt": 1.0,
            "flapping": false,
            "flapping_buffer": 0.0,
            "flapping_current": 0.0,
            "flapping_index": 0.0,
            "flapping_last_change": 0.0,
            "force_next_check": true,
            "force_next_notification": false,
            "last_check_result": null,
            "last_hard_state_change": 1542966756.142521,
            "last_hard_state_raw": 3.0,
            "last_reachable": true,
            "last_state_change": 1542966756.142521,
            "last_state_critical": 0.0,
            "last_state_ok": 0.0,
            "last_state_raw": 3.0,
            "last_state_type": 0.0,
            "last_state_unknown": 0.0,
            "last_state_unreachable": 0.0,
            "last_state_warning": 0.0,
            "next_check": 1545079176.70726,
            "state_raw": 3.0,
            "state_type": 0.0,
            "type": "Service",
            "version": 0.0
        }
    }
```
there is no `last_check_result` and all the other timestamps are not in the future. still the check stays pending, even if I press on "check now" in icingaweb2. also `last_hard_state_change` and `last_state_change` seem to be the restart time icinga2. so why would icinga2 not accept a check result for this service?

**Comment by dgoetz** [comment]
I am facing the same problem here (4 server have been one year in the future). Just to confirm my analyses so far, can you enable the debuglog on the agent and watch it for the execution of the service?
In my case I can see the check execution on the agent, but the master is not updated.

From the state file of the master:
```
    {
        "name": "server.example.com!disk /srv",
        "type": "Service",
        "update": {
            "acknowledgement": 0.0,
            "acknowledgement_expiry": 0.0,
            "check_attempt": 1.0,
            "flapping": false,
            "flapping_buffer": 0.0,
            "flapping_current": 0.0,
            "flapping_index": 7.0,
            "flapping_last_change": 0.0,
            "force_next_check": false,
            "force_next_notification": false,
            "last_check_result": {
                "active": true,
                "check_source": "server.example.com",
                "command": [
                    "/usr/lib64/nagios/plugins/check_disk",
                    "-c",
                    "200",
                    "-w",
                    "500",
                    "-E",
                    "-X",
                    "\"none\", \"sysfs\", \"proc\", \"devtmpfs\", \"devfs\", \"mtmfs\", \"tracefs\", \"cgroup\", \"fuse.gvfsd-fuse\"",
                    "-m",
                    "-p",
                    "/srv"
                ],
                "execution_end": 1578389420.256273,
                "execution_start": 1578389420.25378,
                "exit_status": 0.0,
                "output": "DISK OK - free space: /srv 13265 MB (99.75% inode=100%);",
                "performance_data": [
                    "/srv=32MB;12798;13098;0;13298"
                ],
                "schedule_end": 1578389420.256329,
                "schedule_start": 1578389420.256329,
                "state": 0.0,
                "ttl": 0.0,
                "type": "CheckResult",
                "vars_after": {
                    "attempt": 1.0,
                    "reachable": true,
                    "state": 0.0,
                    "state_type": 1.0
                },
                "vars_before": {
                    "attempt": 1.0,
                    "state": 0.0,
                    "state_type": 1.0
                }
            },
            "last_hard_state_change": 1545404407.004464,
            "last_hard_state_raw": 0.0,
            "last_reachable": true,
            "last_state_change": 1545404407.004464,
            "last_state_critical": 0.0,
            "last_state_ok": 1547026726.635411,
            "last_state_raw": 0.0,
            "last_state_type": 1.0,
            "last_state_unknown": 1545404378.190197,
            "last_state_unreachable": 1545404378.190198,
            "last_state_warning": 0.0,
            "next_check": 1547125547.517258,
            "state_raw": 0.0,
            "state_type": 1.0,
            "type": "Service",
            "version": 0.0
        }
    },
```

On the agent in the debug.log:
```
[2019-01-10 13:53:04 +0100] notice/Process: Running command '/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv': PID 17476
[2019-01-10 13:53:04 +0100] notice/Process: PID 17476 ('/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv') terminated with exit code 0
[2019-01-10 13:53:47 +0100] notice/Process: Running command '/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv': PID 17662
[2019-01-10 13:53:47 +0100] notice/Process: PID 17662 ('/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv') terminated with exit code 0
[2019-01-10 13:55:17 +0100] notice/Process: Running command '/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv': PID 18485
[2019-01-10 13:55:17 +0100] notice/Process: PID 18485 ('/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv') terminated with exit code 0
[2019-01-10 13:56:47 +0100] notice/Process: Running command '/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv': PID 19305
[2019-01-10 13:56:47 +0100] notice/Process: PID 19305 ('/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv') terminated with exit code 0
[2019-01-10 13:58:17 +0100] notice/Process: Running command '/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv': PID 20131
[2019-01-10 13:58:17 +0100] notice/Process: PID 20131 ('/usr/lib64/nagios/plugins/check_disk' '-c' '200' '-w' '500' '-E' '-X' '"none", "sysfs", "proc", "devtmpfs", "devfs", "mtmfs", "tracefs", "cgroup", "fuse.gvfsd-fuse"' '-m' '-p' '/srv') terminated with exit code 0
```

**Comment by dgoetz** [comment]
I think I found the corresponding line of code, if I am correct and checks are executed on the agent but not updated on the master.

https://github.com/Icinga/icinga2/blob/master/lib/icinga/checkable-check.cpp#L171
```
/* Ignore check results older than the current one. */
if (old_cr && cr->GetExecutionStart() < old_cr->GetExecutionStart())
    return;
```

I am not sure if ignoring "older" check results makes sense at all (perhaps only in the case the system has actual results and gets the replay log from a different system) as a now executed check should always be "fresher" than a checkresult created before but perhaps in the future.

**Comment by dnsmichi** [comment]
Removing this line will cause weird behavior in all your external interfaces, and the cluster itself. You want to reset the scheduling queue on the client/satellite itself, not the processing of check results.

**Comment by widhalmt** [comment]
So should it be sufficient to stop the agent/satellite, remove the statefile *there* and just start icinga 2 again?

You're out of luck if the "satellite" is your master, though.

**Comment by Napsty** [comment]
Please note that in my setup there were no agents/satellites involved.

**Comment by Thomas-Gelf** [comment]
I fully back what @marcofl wrote in his initial comment. In my believes we should try to not require manual intervention at all cost. Time traveling happens, caused by configuration errors, hardware, virtualization issues. As a monitoring system, we must be able to detect such problems, to report them, and to report when things went back to normal.

We should not preach to fix things by messing around in `/var/lib/icinga2`. Especially in larger environments with thousands of servers monitoring people often do not even have access to all monitored systems.

**Comment by widhalmt** [comment]
/ref/NC/595861

**Comment by lazyfrosch** [comment]
This is a problem Nagios/Icinga1 also ever had, whenever next_check is set to a date in the future.

Question is, should we change the behavior that next_check can be further away than check_interval dictates? If so, we should reset any next_check that is too far in the future...

**Comment by widhalmt** [comment]
I really like the idea of planning the next check far into the future. An alternative could be to set an end date to "disable active check" just like with acknowledgements.

Could we "just" have it overwrite the next time when to check, when we use "check now" or reschedule the check?

Additionally a predefined Dashboard like "overdue" would be nice for "far in the future"

**Comment by Thomas-Gelf** [comment]
> I really like the idea of planning the next check far into the future

Me too. Having next_check far in the future is absolutely valid.

> Could we "just" have it overwrite the next time when to check, when we use "check now" or reschedule the check?

That should be the expected behavior. The moment I reschedule the check, for example by clicking "check now" this must just work.

@dnsmichi: could you eventually bring some light into this issue? Is the problem "just" the `next_check` field, or are problems with our timestamp-based cluster protocol involved?

**Comment by Crunsher** [comment]
After some time traveling, I managed to reproduce this and hunt down the culprit to:
```
	/* Ignore check results older than the current one. */
	if (old_cr && cr->GetExecutionStart() < old_cr->GetExecutionStart())
		return;
```
In checkable-check.cpp

The question is now what's the correct way to solve this?

Option A: Remove this check and accept whatever checkresult we get.
This safety was added with b887b55 years ago, I am do not know in which case we can expect checkresults out of order. Maybe you can chime in here.

Option B: Still accept checkresult if the old one looks fishy
If the old check result is from the future, something *probably* went wrong and we can use the new one without worry.

**Comment by lippserd** [comment]
I opt for option B.

**Comment by Crunsher** [comment]
PR with a fix implementing **OPTION B** at #6894
